Raxml

mikerobeson opened the pull request:
Added basic [RAxML](https://sco.h-its.org/exelixis/web/software/raxml/index.html) functionality to q2-phylogeny. 

1. Added single-thread and multi-threaded executable options for RAxML.
2. Added RAxML-specific tests along with their associated FASTA and Tree test files.
3. Added RAxML citation.
4. Added dependency and citations for future [IQ-TREE](http://www.iqtree.org/) plugin development.

thermokarst commented on ci/recipe/meta.yaml:
```diff
@@ -24,6 +24,8 @@ requirements:
     - q2-types {{ release }}.*
     - fasttree
+    - raxml
+    - iqtree
```
Where is `iqtree` being used in this PR? Is it a dependency of `raxml`?

thermokarst replied:
Sorry, I misread your PR submission text above. How about we coordinate adding that dependency later, after we get this raxml method in?

mikerobeson replied:
Okay will do. I'll remove the iqtree dependency.

thermokarst commented on ci/recipe/meta.yaml:
```diff
@@ -24,6 +24,8 @@ requirements:
     - qiime2 {{ release }}.*
     - q2-types {{ release }}.*
     - fasttree
+    - raxml
```
Looks like you might need to `merge` or `rebase` your branch using the latest master from https://github.com/qiime2/q2-phylogeny.

mikerobeson replied:
Right! I'll have to figure this out and get back to you.

mikerobeson replied:
Okay, I think I've addressed the conflicts. :-)

thermokarst commented on q2_phylogeny/_raxml.py:
```diff
@@ -0,0 +1,55 @@
+from q2_types.tree import NewickFormat
+
+
+def run_command(cmd, verbose=True, env=None):
```
This command is already defined in `_fasttree.py` - how about we delete this duplicate function and instead import the command:

```python
from q2_phylogeny._fasttree import run_command
```

Does that make sense?

mikerobeson replied:
The run command in `_raxml.py` is different than that in `_fasttree.py`. Mainly, there is not standard input / output like FastTree.

thermokarst commented on q2_phylogeny/citations.bib:
```diff
@@ -10,3 +10,38 @@ @article{price2010fasttree
+URL = {http://dx.doi.org/10.1093/bioinformatics/btu033},
+}
+
+@article{Nguyen2015iqtree,
```
Please see my question above about iqtree.

mikerobeson replied:
Yarp. I'll remove the iqtree citations for this. :-)

thermokarst commented:
BTW, if you want to run the tests and linting locally, you can do the following (in your dev env):

```bash
# one-time, just to install deps into your env
pip install -q flake8 coveralls
pip install -q https://github.com/qiime2/q2lint/archive/master.zip
```

```bash
make lint
make test
```

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,33 @@
+                      'reconstruction.'),
+    },
+    parameter_descriptions={
+        'n_threads': ('The number of threads. Using more than one thread '
```
How about: `The number of threads to use for multithreaded processing. Using more than...`?

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,33 @@
+        'n_threads': ('The number of threads. Using more than one thread '
+                      'will enable the PTHREADS version of RAxML'),
+        'substitution_model': ('Model of Nucleotide Substitution'),
+        'seed': ('Random number seed for the parsimony starting tree.'
```
Missing a space at the end of the line (`starting tree. `). These ones are super tricky, and are normally only caught when looking at the rendered help text via the cli interface (or the Artifact API docstring).

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,33 @@
+                      'will enable the PTHREADS version of RAxML'),
+        'substitution_model': ('Model of Nucleotide Substitution'),
+        'seed': ('Random number seed for the parsimony starting tree.'
+                 'This allows you to reproduce tree results.'
```
Same as above (missing space).

thermokarst commented on q2_phylogeny/tests/test_raxml.py:
```diff
@@ -0,0 +1,186 @@
+    package = 'q2_phylogeny.tests'
+
+    def test_raxml(self):
+        """Test that output tree is made.
```
Thanks for documenting these tests! I haven't seen docstrings used before for this before (as opposed to a plain-old-comment, `#`-style), but am open either way. I want to hear from @ebolyen though, in case this might have some undesired consequence later on.

mikerobeson replied:
Ahh, good point. I did not notice the lack of docstrings elsewhere. I have no problem using regular comments.

thermokarst commented on q2_phylogeny/tests/test_raxml.py:
```diff
@@ -0,0 +1,186 @@
+                              'GCA000686145', 'GCA001950115', 'GCA001971985',
+                              'GCA900007555']))
+
+    def test_raxml_underscore_ids(self):
```
❤️

thermokarst commented on setup.py:
```diff
@@ -14,7 +14,7 @@
     cmdclass=versioneer.get_cmdclass(),
     packages=find_packages(),
-    author="Greg Caporaso",
+    author="Greg Caporaso and Michael Robeson",
```
👍 --- maybe we can think about extending plugin registration to support per-method authors some day in the future. Either way, not sure if we need to also update the `author_email` below - @gregcaporaso & @ebolyen, thoughts?

thermokarst commented on q2_phylogeny/__init__.py:
```diff
@@ -8,10 +8,11 @@
 del get_versions
 
-__all__ = ["midpoint_root", "fasttree", "filter_table"]
+__all__ = ["midpoint_root", "fasttree", "raxml", "filter_table"]
```
💯 - I always forget to do this, myself.

thermokarst commented on q2_phylogeny/_raxml.py:
```diff
@@ -0,0 +1,58 @@
+        run_command(cmd)
+
+        tree_tmp_fp = os.path.join(temp_dir, 'RAxML_bestTree.%s' % runname)
+        os.rename(tree_tmp_fp, str(result))
```
👍

thermokarst commented on q2_phylogeny/_raxml.py:
```diff
@@ -0,0 +1,58 @@
+                '-s', aligned_fp,
+                '-w', temp_dir,
+                '-n', runname]
+        run_command(cmd)
```
I haven't taken a deep dive on the raxml docs - are there any other parameters that might be important to expose to q2-users?

mikerobeson replied:
There. Are. So. Many. Parameters. That is, RAxML uses most of the upper and lower cases of the alphabet for the various flags. Many of which may not apply for our use cases. I decided to keep some parity with what was available in QIIME 1 and add a few more items. I think I will add the rapid bootstrap option. This way users, if interested, can look at the support values of the splits in their tree. But if there are a few other options we should make available, I'd be happy to consider them.

mikerobeson replied:
I forgot to ask, how to add a citation only when a particular flag is enabled? For example, if I add the ability to perform rapid bootstrapping, then I would like to have that specific citation added to the provenance only if that flag was enabled. I remember discussing something along these lines at the meeting but I am unable to find an example of this.

ebolyen replied:
@mikerobeson, that currently isn't possible, a potential future plan would be to create "hidden actions" which are then invoked in a pipeline based on some condition (like that flag). I wonder if that's the right approach for something like this however.

mikerobeson replied:
@ebolyen okay. To get around this issue in the short term I can setup a separate function called `raxml-bootstrap`. Then I can simply place all relevant citations and extra parameters there. What do you think? Another option is to hold the bootsrap feature for a later pull request?

ebolyen replied:
Either works fine I think, and actually an action like raxml-bootstrap might actually make more sense on its own anyhow :+1:, for example you could indicate that the output includes support values on the tree which would be useful to know.

mikerobeson replied:
I've added `raxml-rapid-bootstrap` as a separate function. I have a couple more simple tests to work out, but the plugin is functional. I do have one question though: how can I properly test if the number of bootstrap replicates I requested is actually being performed / passed in? 

That is, RAxML will write out a single file in which all the individual trees produced for each of these bootstrap replicates resides. For now, this collection of individual bootstrap trees is not retained in the output of `raxml-rapid-bootstrap`, only the final tree with the bootstrap supports is. I am trying to determine the best way to keep this file of trees, or at least return it for testing purposes. I figure, the simplest test would be to count the number of trees written to this file and make sure that is the same count as the number of bootstrap replicates I've requested.

Any suggestions?

thermokarst commented on q2_phylogeny/citations.bib:
```diff
@@ -10,3 +10,16 @@ @article{price2010fasttree
   doi={10.1371/journal.pone.0009490}
 }
+
+@article{Stamatakis2014raxml,
```
🎉

thermokarst commented on q2_phylogeny/citations.bib:
```diff
@@ -10,3 +10,16 @@ @article{price2010fasttree
 }
+
+@article{Stamatakis2014raxml,
+author = {Stamatakis, Alexandros},
```
Sorry, this is a little nitpicky, but do you think you could match the line indentation for 15-23 to match the citation entries above?

thermokarst replied:
Excellent point @ebolyen - I think I was coming from less of a "style" perspective and more of a "Bibtex PTSD" perspective - playing with poorly behaved parsers in the past has scarred me. If you think this is purely a style issue, then I am all on board with leaving things as-is (here, and in general)!

ebolyen replied:
Ditto, we'll get the worlds longest stacktrace (like 3 nested raise-froms) when something is actually wrong. Otherwise, this will all get massaged into a standard indentation, curly-brace-quoted, and latex-decoded format.

mikerobeson replied:
Ahh, no problem. I am surprised my indentation OCD failed me in a big way here. :-)

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,72 @@
+        'seed': ('Random number seed for the parsimony starting tree. '
+                 'This allows you to reproduce tree results. '
+                 'If not supplied then one will be randomly chosen. '),
+        'rapid_bootstrap_seed': ('Specify a random seed for rapid '
```
Can you tighten up the line breaks on for this description, it is a little hard to read in the source.

e.g.

```python
'rapid_bootstrap_seed': ('Specify a random seed for rapid '
                         'bootstrapping. This allows you to reproduce '
                         'rapid bootstrap results. If not supplied '
                         'then one will be randomly chosen.'),
```

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,72 @@
+                                 'If not supplied then one will be randomly '
+                                 'chosen.'),
+        'bootstrap_replicates': ('The number of bootstrap searches to '
+                                 'perform. ')},
```
Extra space at the end of the sentence.

thermokarst commented on q2_phylogeny/plugin_setup.py:
```diff
@@ -60,6 +62,72 @@
+                                 'chosen.'),
+        'bootstrap_replicates': ('The number of bootstrap searches to '
+                                 'perform. ')},
+    output_descriptions={'tree': 'The resulting phylogenetic tree. '},
```
Extra space at the end of the sentence.

thermokarst commented on q2_phylogeny/tests/test_raxml.py:
```diff
@@ -0,0 +1,284 @@
+        # This test is comparing an ordered series of tip-to-tip distances
+        # to a tree output from a manual run of the default command:
+        # raxmlHPC -m GTRGAMMA -p 1723 -s aligned-dna-sequences-3.fasta -n q2
+        # NOTE: I cleanly rounded the tip-to-tip dists (i.e. `%.4f`) as RAxML
```
👍

mikerobeson commented:
@thermokarst Thanks for those last set of comments! I've added `_build_rapid_bootstrap_command` and its associated test case. Then I added an additional flag to `raxml`. I am set to go with the merge if these additions check out. 😃

thermokarst commented:
Thanks @mikerobeson!

A handful of upstream packages recently changed, which has [created a bit of cleanup work](https://busywork.qiime2.org/teams/main/pipelines/master-distribution?groups=all) for us on our build system. I am waiting to do the final review and merge after we address those (unrelated problems). Just wanted to give you an update on the hold-up here! Thanks for being patient!

thermokarst commented:
@mikerobeson, now that this is merged, any thoughts on contributing a [Community Tutorial](https://forum.qiime2.org/c/community-contributions/community-tutorials)?

mikerobeson commented:
I'd be happy to contribute to a tutorial!
